Sources/SwiftProtobuf/BinaryDecoder.swift

```diff
@@ -1007,11 +1007,22 @@ internal struct BinaryDecoder: Decoder {
 
     internal mutating func decodeExtensionField(values: inout ExtensionFieldValueSet, messageType: Message.Type, fieldNumber: Int) throws {
         if let ext = extensions?[messageType, fieldNumber] {
-            var fieldValue = values[fieldNumber] ?? ext._protobuf_newField()
             assert(!consumed)
-            try fieldValue.decodeExtensionField(decoder: &self)
+            var fieldValue = values[fieldNumber]
+            if fieldValue != nil {
+                try fieldValue!.decodeExtensionField(decoder: &self)
+            } else {
+                fieldValue = try ext._protobuf_newField(decoder: &self)
+            }
             if consumed {
-              values[fieldNumber] = fieldValue
+                if fieldValue != nil {
+                    values[fieldNumber] = fieldValue
+                } else {
+                    // Really things should never get here as the required data
+                    // should always be available or the decode of the extension
+                    // would throw.
+                    throw BinaryDecodingError.truncated
```

Comment:
Yeah, we probably should—that way if this line ever *does* get executed, we have a better idea of what might have gone wrong without having to remember all of these details.

Actually... since Swift errors don't come with backtraces, figuring out that it was *this* truncated error vs. some other truncated error gets a little annoying (you'd have to place a breakpoint and wait for things to go south). I wonder if we should introduce a new error case, perhaps with a string reason, to handle these kinds of internal/illegal-state errors that we don't want to bring down the whole world?